### PR TITLE
feat: Implement cert transfer interface

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

**Note**: Depends on [this change](https://github.com/canonical/certificate-transfer-interface/pull/96) in the cert transfer. Should be merged after.

- Implements the cert transfer interface in the lego base lib.
- Adds Tests
- The CA certificates are fetched from the databag of the certificates relation and moved to the databag of the cert transfer relation

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I Have bumped the version of the library

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -547,3 +556,59 @@ def test_given_cmd_when_app_environment_variables_set_then_command_executed_with
                 "NO_PROXY": "No proxy",
             },
         )
+
+    @patch(
+        f"{CERT_TRANSFER_LIB_PATH}.CertificateTransferProvides.add_certificates",
+    )
+    def test_given_cert_transfer_relation_not_created_then_ca_certificates_not_added_in_relation_data(  # noqa: E501
```
Comment (kayra1) [inline tests/unit/test_charm.py:563]: do we need to patch this? we can just look at the provider databag and see that its empty

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -547,3 +556,59 @@ def test_given_cmd_when_app_environment_variables_set_then_command_executed_with
...
+        self.harness.charm.on.config_changed.emit()
+        mock_add_certificates.assert_not_called()
+
+    @patch(
+        f"{TLS_LIB_PATH}.TLSCertificatesProvidesV3.get_provider_certificates",
+    )
+    @patch(
+        f"{CERT_TRANSFER_LIB_PATH}.CertificateTransferProvides.add_certificates",
+    )
+    def test_given_cert_transfer_relation_and_ca_certificates_then_ca_certificates_added_in_relation_data(  # noqa: E501
```
Comment (kayra1) [inline tests/unit/test_charm.py:586]: same comment as above

**Review by kayra1 — CHANGES_REQUESTED** [review]
2 small suggestions. Great job.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py:563]: I'm not sure how to do that. If there is no relation, there is no databag, I can't do `harnsess.get_relation_data(relation_id, app_name)`.
I fixed it for the other comment however.

File: tests/unit/test_charm.py (hunk 2 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py]: Done.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (kayra1) [reply tests/unit/test_charm.py:563]: in that case wouldn't the charm just raise an error? So we don't need to do either, we can just wait for the test to raise if it breaks.

Comment (saltiyazan) [reply tests/unit/test_charm.py:563]: No, the charm will check whether the relation was created and it won't crash or raise.

Comment (kayra1) [reply tests/unit/test_charm.py:563]: i mean if this feature was bugged and the charm attempted to add ca certificates to a relation that is not created. In that case nothing will happen?

Comment (saltiyazan) [reply tests/unit/test_charm.py:563]: That's more on the lib to check, the charm will not directly interact with relation data.

**Review by kayra1 — APPROVED** [review]

**Pull request merged**